Add --interval for container installation

addshore opened the pull request:
This mimics the interval options that currently exist
within artemis.
You can pass a time period, which jaguar will use to ensure
that your container is running. If it isn't running its started
again.

addshore commented:
`./build/jag container install  error ~/dev/lb/io/toit/src/error.toit --interval 4s`

Would result in

```
~/dev/github/toitlang/jaguar (main) ./build/jag simulate
[jaguar.http] INFO: running Jaguar device 'silly-bench' (id: 'a0e467c9-a928-4969-be07-19e05ff2835b') on 'http://172.21.230.90:45063'
[jaguar] INFO: container 'error' installed and started with {jag.interval: 4s} (interval: 4s)
This will error in 5s

******************************************************************************
EXCEPTION error.
This is an error message
  0: main                      /home/adam/dev/lb/io/toit/src/error.toit:4:3
******************************************************************************

[jaguar] ERROR: container 'error' stopped - exit code 1
[jaguar] INFO: restarting container 'error' (interval check)
[jaguar] INFO: container 'error' restarted with {jag.interval: 4s} (interval: 4s)
This will error in 5s

******************************************************************************
EXCEPTION error.
This is an error message
  0: main                      /home/adam/dev/lb/io/toit/src/error.toit:4:3
******************************************************************************

[jaguar] ERROR: container 'error' stopped - exit code 1
[jaguar] INFO: restarting container 'error' (interval check)
[jaguar] INFO: container 'error' restarted with {jag.interval: 4s} (interval: 4s)
This will error in 5s

******************************************************************************
EXCEPTION error.
This is an error message
  0: main                      /home/adam/dev/lb/io/toit/src/error.toit:4:3
******************************************************************************

[jaguar] ERROR: container 'error' stopped - exit code 1
^C
```

floitsch commented:
> Another issue currently seems to be that if you install a container with an interval, and then again without one, the monitor doesn't get cleaned up.
> 
> So your second container gets restarted even if you were not expecting that to happen

good point.

addshore commented:
So, I'm suite happy with this solution now, but still have some questions.

I see lots of conditionals for `if name`, however the jag CLI at least doesn't allow a container to be installed without a name...

Is a nameless container something I should be testing against? (as this first counter based implementation likely has a couple of issues with nameless containers to iron out)

floitsch commented on src/container_registry.toit:
```diff
@@ -18,10 +18,12 @@ class ContainerRegistry:
   id-by-name_         / Map ::= {:}  // Map<string, uuid.Uuid>
   name-by-id_         / Map ::= {:}  // Map<uuid.Uuid, string>
   entry-by-id-string_ / Map ::= {:}  // Map<string, List>
+  run-counters_       / Map ::= {:}  // Map<string, int> - container name to run counter
```
```suggestion
  /** The current revision of the container. Reset to 0 at every boot. */
  revisions_       / Map ::= {:}  // Map<string, int>
```

floitsch commented on src/container_registry.toit:
```diff
@@ -87,6 +89,15 @@ class ContainerRegistry:
     store_
```
Add a comment, explaining that we don't remove the entry from the revisions, so that we guarantee that a newer revision of a program still has a newer revision-number, even if it was uninstalled at some point.

floitsch commented on src/container_registry.toit:
```diff
@@ -87,6 +89,15 @@ class ContainerRegistry:
+    run-counters_[name] = new-count
+    return new-count
+
+  get-run-counter name/string -> int:
```
```suggestion
  revision name/string -> int:
```

floitsch commented on src/jaguar.toit:
```diff
@@ -317,6 +332,16 @@ start-image_ -> bool
+      remaining-us := interval.in-us - (Time.monotonic-us - start-time)
+      scheduled-callbacks.add (Duration --us=remaining-us+1) --callback=::
+        current-run-number := registry_.get-run-counter name
+        if current-run-number == run-number and registry_.id-by-name_.contains name:
```
Add a function to the registry_ to support `contains name`. No need to reach into the privates.

floitsch reviewed (commented):
LGTM with easy comments.

addshore commented:
Cleaned up some stuff around names given what you said, and changed the order slightly.
And added a throw for name and interval (which shouldnt happen)

addshore commented:
Infact, I have removed the throw, as due to the condition changes it would never occur anyway.
Lets save defining the string? :)

addshore commented on cmd/jag/commands/container.go:
```diff
@@ -140,6 +160,9 @@ func ContainerInstallCmd() *cobra.Command {
 	cmd.Flags().String("assets", "", "attach assets to the container")
 	cmd.Flags().IntP("optimization-level", "O", -1, "optimization level")
+	cmd.Flags().String("interval", "", "check interval for container monitoring (e.g., '30s', '5m', '1h'). "+
+		"When specified, Jaguar will periodically check if the container is running and restart it if it has stopped. "+
```
Tweaked the text (will commit soon), and this is what it currently looks like

![image](https://github.com/user-attachments/assets/417a4605-ef2c-4002-8384-d19554610ae7)

or wrapped

![image](https://github.com/user-attachments/assets/be87755e-f022-48c7-9fa8-f80ac53b3102)

Which is probably fine, as other text already wraps in this way? (see watch below)

![image](https://github.com/user-attachments/assets/d4061ce2-a393-4bca-9dfd-d537d26852ab)

floitsch commented on cmd/jag/commands/container.go:
```diff
@@ -140,6 +160,8 @@ func ContainerInstallCmd() *cobra.Command {
 	cmd.Flags().StringArrayP("define", "D", nil, "define settings to control container on device")
 	cmd.Flags().String("assets", "", "attach assets to the container")
 	cmd.Flags().IntP("optimization-level", "O", -1, "optimization level")
+	cmd.Flags().String("interval", "", "check interval for container starts (e.g., '30s', '5m', '1h'). "+
+		"When specified, Jaguar will start the container at the specified interval if it has previously exited.")
```
I think the interval help is too complicated to fit nicely in the short help.

I would add something like the following to the big help above:
```
If an interval is given, the container is started at the given frequency. If the container is
still running, then all subsequent starts are simply delayed. If a container should run
continuously, an interval of `1s` is a good choice. If the container runs more than 1 second
it will be restarted automatically, but if it crashes immediately, then it won't be restarted
for 1 second, giving the system more room to recover.
```

```suggestion
	cmd.Flags().String("interval", "", "interval for container starts (like '30s', '5m', or '1h').")
```

addshore replied:
Done in https://github.com/toitlang/jaguar/pull/624/commits/9f21871beeffd0bfacf76ab3820f1ab1d6a530f0

floitsch commented on cmd/jag/commands/run.go:
```diff
@@ -356,6 +356,17 @@ func sendCodeFromFile(
+					}
+					headersMap[JaguarContainerIntervalHeader] = converted
+				default:
+					return fmt.Errorf("jag.interval must be a duration string (e.g., '30s', '5m', '1h')")
```
This way it's similar to the `jag.timeout`.
```suggestion
					return fmt.Errorf("cannot parse jag.interval ('%s') as a duration", converted)
```

floitsch commented on src/container_registry.toit:
```diff
@@ -47,6 +50,8 @@ class ContainerRegistry:
       id-by-name_[name] = id
       name-by-id_[id] = name
       entry-by-id-string_[id-as-string] = [name, defines, id]
+      if name:
+        revisions_[name] = 0
```
```suggestion
      if name: revisions_[name] = 0
```

floitsch commented on src/container_registry.toit:
```diff
@@ -76,6 +81,8 @@ class ContainerRegistry:
     name-by-id_[id] = name
     entry-by-id-string_["$id"] = [name, defines, id]
     store_
+    if name:
+      revisions_[name] = (revisions_.get name --if-absent=: 0) + 1
```
```suggestion
    if name: revisions_.update name --if-absent=0: it + 1
```

floitsch reviewed (commented):
Nice.
Probably the last round of comments.
